### DISCUSSION TITLE
How to use transaction inside Job OR how to rollback model write operation inside job when it runs out of memory.

### DISCUSSION THREAD

**Discussion by Himakar-PV** [laravel/framework#37357] [Q&A] [answered]
Hi,

I have a job which process below operations:

1. import data, save in db
2. change status of saved records
3. dispatch event with saved records for further operations.

At point 2. it runs out of memory and process get killed.
So I want to use a transaction here so that if point 2 not executed completely I want to rollback point 1 operation.

Currently I am doing this:

````
public function handle()
    {
        
        $model->beginTransaction();

        $importer->import($model); // operation 1

        $convertedItem = $service->digitize($model); // operation 2

        $bill = $service->linkBill($convertedItem); // operation 3

        $model->commit();

        $dispatcher->dispatch(new Event($model, $convertedItem, $bill)); // dispatching event
    }

    /**
     * Handling Failure exception
     *
     * @param \Throwable $exception
     */
    public function failed(\Throwable $exception)
    {
        Log::error('Exception at job failure :  {exception}', [
            'exception' => $exception->getMessage()
        ]);
    }
````

If no memory leak then everything works fine, if memory leak happened at operation 2 then I can not know it and rollback the model ($model->rollback()).

How to achieve this? Currently it is just leaving open transactions as they are not committed when memory leak happened.

**Comment by denisdulici** [comment]
Unfortunately, that's out of the scope of transactions as PHP is terminated due to the memory leak.

**Reply by Himakar-PV** [reply]
Thank you for the reply. 🙏 

I found a way to do it. In fact it is not out of scope.

Laravel queue worker checks this and log on the terminal when memory leak happens and the same can be fired as event and can listen in job but such event is not made by laravel.

Simple way to handle this scenario is, inside job handle method can write like this:

```
error_reporting(-1);
        register_shutdown_function(function() use($purchaseOrderItemRepository, $logger) {
            $error = error_get_last();
            if (null !== $error) {
                logger()->info('Queue died with an error : '. json_encode($error));
                // write rollback here...
            }
        });
```

This works perfectly fine.

**Comment by Himakar-PV** [comment, marked as answer]
Any one looking for solution, can use this in job handle method:
```
error_reporting(-1);
        register_shutdown_function(function() use($purchaseOrderItemRepository, $logger) {
            $error = error_get_last();
            if (null !== $error) {
                logger()->info('Queue died with an error : '. json_encode($error));
                // write rollback here...
            }
        });
```

hope laravel will come up with nice way of handling this to get caught in the failed method.

**Reply by denisdulici** [reply]
How come you can `rollback` when PHP is terminated? The purpose of [transactions](https://laravel.com/docs/8.x/database#database-transactions) seems to be catching exceptions. I may be missing something here 😢

**Reply by Himakar-PV** [reply]
Hi bro, we are using Model Transactions, purpose is to hold the write operations into database until transaction is committed. PHP job will be terminated when memory leak. The moment memory leak happened, as per above function it logs the message and model transaction gets rollback. I believe rollback is working because I can log message even after rollback statement.

I also have the same doubt how it even log message when script execution stopped may be at the core level it will have the memory to log and execute this command.

Atleast I can say this that transaction is not committed and the write operation is not performed at that failure.
